File: 백준/Gold/11000. 강의실 배정/강의실 배정.py
import sys
import heapq
n = int(sys.stdin.readline())
times= [list(map(int,sys.stdin.readline().split())) for _ in range(n)]

times.sort()
room= []
heapq.heappush(room,times[0][1])
for i in range(1, n):
    start=times[i][0]  
    if times[i][0] < room[0]  :
        heapq.heappush(room,times[i][1])
    else:
        heapq.heappop(room)
        heapq.heappush(room,times[i][1])
print(len(room))

File: 백준/Gold/11000. 강의실 배정/강의실 배정.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
    
        List<List<Integer>> times = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            List<String> line =Arrays.asList(br.readLine().split(" "));
            List<Integer> integerLine = new ArrayList<>();
            for (String str : line) {
                integerLine.add(Integer.parseInt(str));
            }
            times.add(integerLine);
        }
        Collections.sort(times, (o1, o2)-> o1.get(0).compareTo(o2.get(0)));
        PriorityQueue<Integer> room = new PriorityQueue<>();
        room.add(times.get(0).get(0));

        for(int i=0;i<n;i++){
            int start = times.get(i).get(0);
            if(start < room.peek()){
                room.add(times.get(i).get(1));
            }else{
                room.poll();
                room.add(times.get(i).get(1));
            }
        }
        System.out.println(room.size());
    }
}
